src/core/xds/grpc/xds_cluster_parser.cc

```diff
@@ -641,34 +641,49 @@ absl::StatusOr<std::shared_ptr<const XdsClusterResource>> CdsResourceParse(
-            &telemetry_labels_struct)) {
-      size_t iter = kUpb_Map_Begin;
-      const google_protobuf_Struct_FieldsEntry* fields_entry;
-      while ((fields_entry = google_protobuf_Struct_fields_next(
-                  telemetry_labels_struct, &iter)) != nullptr) {
-        // Adds any entry whose value is a string to telemetry_labels.
-        const google_protobuf_Value* value =
-            google_protobuf_Struct_FieldsEntry_value(fields_entry);
-        if (google_protobuf_Value_has_string_value(value)) {
-          if (UpbStringToAbsl(google_protobuf_Struct_FieldsEntry_key(
-                  fields_entry)) == "service_name") {
-            cds_update->service_telemetry_label = RefCountedStringValue(
-                UpbStringToAbsl(google_protobuf_Value_string_value(value)));
-          } else if (UpbStringToAbsl(google_protobuf_Struct_FieldsEntry_key(
-                         fields_entry)) == "service_namespace") {
-            cds_update->namespace_telemetry_label = RefCountedStringValue(
-                UpbStringToAbsl(google_protobuf_Value_string_value(value)));
-          }
-        }
+    // First, try typed_filter_metadata.
+    size_t iter = kUpb_Map_Begin;
+    const envoy_config_core_v3_Metadata_TypedFilterMetadataEntry* typed_entry;
+    while (
+        (typed_entry = envoy_config_core_v3_Metadata_typed_filter_metadata_next(
+             metadata, &iter)) != nullptr) {
+      absl::string_view key = UpbStringToAbsl(
+          envoy_config_core_v3_Metadata_TypedFilterMetadataEntry_key(
+              typed_entry));
+      ValidationErrors::ScopedField field(
+          &errors, absl::StrCat(".metadata.typed_filter_metadata[", key, "]"));
+      auto extension = ExtractXdsExtension(
+          context,
+          envoy_config_core_v3_Metadata_TypedFilterMetadataEntry_value(
+              typed_entry),
+          &errors);
+      if (!extension.has_value()) continue;
+      // TODO(roth): If we ever need to support another type here, refactor
+      // this into a separate registry.
+      if (extension->type == "extensions.filters.http.gcp_authn.v3.Audience") {
+        // TODO(roth): In a subsequent PR, add parsing here.
+      }
+    }
+    // Then, try filter_metadata.
+    iter = kUpb_Map_Begin;
+    const envoy_config_core_v3_Metadata_FilterMetadataEntry* entry;
+    while ((entry = envoy_config_core_v3_Metadata_filter_metadata_next(
+                metadata, &iter)) != nullptr) {
+      absl::string_view key = UpbStringToAbsl(
+          envoy_config_core_v3_Metadata_FilterMetadataEntry_key(entry));
+      auto json = ParseProtobufStructToJson(
+          context,
+          envoy_config_core_v3_Metadata_FilterMetadataEntry_value(entry));
+      if (!json.ok()) {
+        ValidationErrors::ScopedField field(
+            &errors, absl::StrCat(".metadata.filter_metadata[", key, "]"));
+        errors.AddError(json.status().message());
+      } else if (!cds_update->metadata.contains(key)) {
```

Comment:
is ignoring duplicates the expected behavior? I don't see it mentioned in the gRFC

Reply:
The gRFC says:

"""
The value for a given metadata key will come from only one of the two maps; the value from `filter_metadata` will be used only if the key is not present or is of an unknown protobuf message type in `typed_filter_metadata`.
"""

We check `typed_filter_metadata` first, and we use the entry in `filter_metadata` only if we didn't already get one from `typed_filter_metadata`.